src/stemmablenote.ts

```diff
@@ -178,66 +200,60 @@ export class StemmableNote extends Note {
   }
 
   // Set the stem length to a specific. Will override the default length.
-  setStemLength(height) {
+  setStemLength(height: number): this {
     this.stem_extension_override = height - Stem.HEIGHT;
     return this;
   }
 
   // Get the top and bottom `y` values of the stem.
-  getStemExtents() {
-    return this.stem.getExtents();
+  getStemExtents(): Record<string, number> | undefined {
+    return this.stem?.getExtents();
   }
 
-  // Sets the current note's beam
-  setBeam(beam) {
-    this.beam = beam;
-    return this;
-  }
-
-  // Get the `y` value for the top/bottom modifiers at a specific `textLine`
-  getYForTopText(textLine) {
+  /** Gets the `y` value for the top modifiers at a specific `textLine`. */
+  getYForTopText(textLine: number): number {
     const extents = this.getStemExtents();
     if (this.hasStem()) {
       return Math.min(
-        this.stave.getYForTopText(textLine),
-        extents.topY - this.render_options.annotation_spacing * (textLine + 1)
+        this.stave?.getYForTopText(textLine) || 0,
+        extents?.topY || 0 - this.render_options.annotation_spacing * (textLine + 1)
       );
     } else {
-      return this.stave.getYForTopText(textLine);
+      return this.stave?.getYForTopText(textLine) || 0;
     }
   }
 
-  getYForBottomText(textLine) {
+  /** Gets the `y` value for the bottom modifiers at a specific `textLine`. */
+  getYForBottomText(textLine: number): number {
     const extents = this.getStemExtents();
     if (this.hasStem()) {
       return Math.max(
-        this.stave.getYForTopText(textLine),
-        extents.baseY + this.render_options.annotation_spacing * textLine
+        this.stave?.getYForTopText(textLine) || 0,
```

Comment:
These `|| 0`s are dangerous.

Reply:
|| 0 applied to a number will result in 0 if the number is undefined, null, 0, NaN.
Should I go for ?? 0 that only makes the change is number is undefined or null? Or you prefer a different approach?

Reply:
Yes, I understand, but it's an implementation change and it's hard to reason about what happens. Somtimes a different layer of code deals with undefined/null. Or sometimes, it should just crash (with console error/exception) instead of rendering something incorrect.

This applies to `stave?` too -- it changes the behavior and expectations of the program. Those should stay `stave`. (You can use `stave?` if you see something like `if (stave) ... else undefined` or `a = f ? c : null`

Reply:
@0xfe explanation:
- `stave` is a member of `Element`
- `stave` is not assigned a value in the `constructor` of `Element`. For this reason it is declared `stave?: Stave;`
- due to the fact that TS knows that `stave` can be `undefined` it forces us to say how do we want to handle that. This is not the case in JS.

The approach that I have selected is:
`this.stave?.getYForTopText(textLine) || 0`
basicly if stave is undefined or getYFortTopText does not return a valid number then 0. I will change it to the mos restrictive 
`this.stave?.getYForTopText(textLine) ?? 0`

Reply:
If it forces to handle that `stave` can be undefined then I would prefer `this.stave?.getYForTopText(textLine) ?? undefined` as opposed `?? 0` because then it fails loudly instead of silently returning 0 with a weird formatting problem that is difficult to diagnose. It's the zero-default which is the problem here.

Alternatively, we can:

- make `stave` non-optional, and initialize to `null`.
- add an assertion before the line, something like, `assertDefined(this.stave)` (we'll have to build some standard assertions.)